aptos-move/framework/aptos-token/sources/bid.move

```diff
@@ -0,0 +1,343 @@
+    /// ensure the coin cannot be withdrawn or transferred during the auction
+    struct Bid<phantom CoinType> has store {
+        bidder: address,
+        coin: Coin<CoinType>,
+        offer_price: u64,
+        listing_id: ID, // ensure the bid is only executed for the chosen listing_id
+    }
+
+    /// store all the bids by the user
+    struct BidRecords<phantom CoinType> has key {
+        records: Table<ID, Bid<CoinType>>,
+        bid_event: EventHandle<BidEvent<CoinType>>,
+        withdraw_bid_event: EventHandle<WithdrawBidEvent<CoinType>>,
+    }
+
+    struct BidEvent<phantom CoinType> has copy, drop, store {
+        offer_price: u64,
+        listing_id: ID,
+        bid_id: ID,
+    }
+
+    struct WithdrawBidEvent<phantom CoinType> has copy, drop, store {
+        bid_id: ID,
+    }
+
+    public fun initialize_bid_records<CoinType>(bidder: &signer)  {
+        let owner_addr = signer::address_of(bidder);
+
+        if (!exists<BidRecords<CoinType>>(owner_addr)) {
+            move_to(
+                bidder,
+                BidRecords<CoinType> {
+                    records: table::new(),
+                    bid_event: account::new_event_handle<BidEvent<CoinType>>(bidder),
+                    withdraw_bid_event: account::new_event_handle<WithdrawBidEvent<CoinType>>(bidder),
+                }
+            );
+        };
+    }
+
+    public fun assert_bid_for_a_listing<CoinType>(
+        token_id: TokenId,
+        offer_price: u64,
+        token_amount: u64,
+        entry: &Listing<CoinType>
+    ) {
+        // validate token_id match
+        assert!(token_id == listing::get_listing_token_id(entry), ETOKEN_ID_NOT_MATCH);
+        // validate offerred amount and price
+        let listed_amount =  listing::get_listing_token_amount(entry);
+        let min_total = listing::get_listing_min_price(entry) * listed_amount;
+        assert!(offer_price * token_amount >= min_total, ENO_SUFFICIENT_FUND);
+        assert!(token_amount == listed_amount, ETOKEN_AMOUNT_NOT_MATCH);
+    }
+
+    /// withdraw the coin and store them in bid struct and return a global unique bid id
+    public fun bid<CoinType>(
+        bidder: &signer,
+        token_id: TokenId,
+        token_amount:u64,
+        offer_price: u64,
+        entry: &Listing<CoinType>,
+    ): ID acquires BidRecords {
+        initialize_bid_records<CoinType>(bidder);
+        let bidder_address = signer::address_of(bidder);
+        // check the bid is legit for the listing
+        assert_bid_for_a_listing(token_id, offer_price * token_amount, token_amount, entry);
+        // check bidder has sufficient balance
+        assert!(coin::balance<CoinType>(bidder_address) >= offer_price * token_amount, error::invalid_argument(ENO_SUFFICIENT_FUND));
+        // withdraw the coin and store them in escrow to ensure the fund is avaliable until expiration_sec
+        let coin = coin::withdraw<CoinType>(bidder, offer_price * token_amount);
```

Comment:
this `offer_price * token_amount` should probably be commented somewhere besides reading the code.

Reply:
addressed